keanu-project/src/main/java/io/improbable/keanu/util/io/VertexDotLabel.java

```diff
@@ -32,23 +33,33 @@ public void setValue(String value) {
     public String inDotFormat() {
         // Output value if value is set, but also add some descriptive info for non-constant vertices.
         if (!value.isEmpty()) {
-            String dotLabel = vertex.getId().hashCode() + DOT_LABEL_OPENING + value;
-            if (!(vertex instanceof ConstantVertex)) {
-                dotLabel += " (" + getDescriptiveInfo() + ")";
-            }
-            return dotLabel + DOT_LABEL_CLOSING;
+            StringBuilder dotLabel = new StringBuilder();
+            dotLabel.append(vertex.getId().hashCode()).append(DOT_LABEL_OPENING).append(value);
+            getDescriptiveInfoForVertexWithValue().ifPresent(info -> dotLabel.append(" (").append(info).append(")"));
+            return dotLabel.append(DOT_LABEL_CLOSING).toString();
         }
         return vertex.getId().hashCode() + DOT_LABEL_OPENING + getDescriptiveInfo() + DOT_LABEL_CLOSING;
     }
 
+    private Optional<String> getDescriptiveInfoForVertexWithValue() {
+        if (!vertexLabel.isEmpty()) {
+            return Optional.of(vertexLabel);
+        }
+        if (!(vertex instanceof ConstantVertex)) {
+            return Optional.of(getAnnotationIfPresentElseSimpleName());
+        }
+        return Optional.empty();
+    }
+
     private String getDescriptiveInfo() {
         if (!vertexLabel.isEmpty()) {
             return vertexLabel;
         }
-        if (!annotation.isEmpty()) {
-            return annotation;
-        }
-        return vertex.getClass().getSimpleName();
+        return getAnnotationIfPresentElseSimpleName();
+    }
```

Comment:
We could combine `getDescriptiveInfo` and `getDescriptiveInfoForVertexWithValue` into one method like this. I think this is easier (although still not easy!) to understand overall since it puts all the branching into one place.

    private Optional<String> getDescriptiveInfo() {
        if (!vertexLabel.isEmpty()) {
            return Optional.of(vertexLabel);
        } else if (value.isEmpty() || !(vertex instanceof ConstantVertex)) {
            return Optional.of(getAnnotationIfPresentElseSimpleName());
        } else {
            return Optional.empty();
        }
    }

Reply:
Thank you! This is nicer